internal/persistence/postgres/pool.go

```diff
@@ -0,0 +1,154 @@
+package postgres
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"go.uber.org/zap"
+)
+
+var pools = map[string]PoolOpenFunc{}
+
+// Pool is the interface of the pgxpool.Pool type, used by the postgres
+// persistence layer to query the database. A custom pool type can be
+// registered using `RegisterPool` if it implements this interface.
+type Pool interface {
+	Close()
+	Acquire(ctx context.Context) (*pgxpool.Conn, error)
+	AcquireFunc(ctx context.Context, f func(*pgxpool.Conn) error) error
+	AcquireAllIdle(ctx context.Context) []*pgxpool.Conn
+	Config() *pgxpool.Config
+	Stat() *pgxpool.Stat
+	Exec(ctx context.Context, sql string, arguments ...interface{}) (pgconn.CommandTag, error)
+	Query(ctx context.Context, sql string, args ...interface{}) (pgx.Rows, error)
+	QueryRow(ctx context.Context, sql string, args ...interface{}) pgx.Row
+	QueryFunc(ctx context.Context, sql string, args []interface{}, scans []interface{},
+		f func(pgx.QueryFuncRow) error) (pgconn.CommandTag, error)
+	SendBatch(ctx context.Context, b *pgx.Batch) pgx.BatchResults
+	Begin(ctx context.Context) (pgx.Tx, error)
+	BeginTx(ctx context.Context, txOptions pgx.TxOptions) (pgx.Tx, error)
+	BeginFunc(ctx context.Context, f func(pgx.Tx) error) error
+	BeginTxFunc(ctx context.Context, txOptions pgx.TxOptions, f func(pgx.Tx) error) error
+	CopyFrom(ctx context.Context, tableName pgx.Identifier, columnNames []string, rowSrc pgx.CopyFromSource) (int64, error)
+	Ping(ctx context.Context) error
+}
+
+// PoolOpenFunc defines a function that can instantiate a Pool instance with the given options.
+type PoolOpenFunc func(opts Opts, logger *zap.Logger) (Pool, error)
+
+type PoolOpts struct {
+	// Optional The name of the postgres pool implementation to use.
+	// By default, only `pgx` is available.
+	Name string
+
+	// MaxConns is the maximum size of the pool. Default to `persistence.DefaultMaxConn`.
+	MaxConns int32
+
+	// MinConns is the minimum size of the pool. Default to `persistence.DefaultMinConn`.
+	MinConns int32
+
+	// MaxConnLifetime is the duration since creation after which a connection will
+	// be automatically closed. Default to `persistence.DefaultMaxConnLifetime`.
+	MaxConnLifetime time.Duration
+
+	// MaxConnIdleTime is the duration after which an idle connection will be automatically
+	// closed by the health check. Default to `persistence.DefaultMaxConnIdleTime`.
+	MaxConnIdleTime time.Duration
+
+	// HealthCheckPeriod is the duration between checks of the health of
+	// idle connections. Default to `persistence.DefaultHealthCheckPeriod`.
+	HealthCheckPeriod time.Duration
+
+	// ReadOnly indicates the pool will be used for read only operations.
+	ReadOnly bool
```

Comment:
How will this be used, as it's currently unused?

Reply:
currently, this is used in the aurora impl to determine what kind of healthcheck validator is required (read vs write). However, we could remove that if we had a generic `PoolOpts.params` that we can pass thru.

Reply:
I am not sure about the value of the ReadOnly parameter as well. Perhaps this is something to talk about internally?

Reply:
sure Rajan. Here is how the ro pool is currently created: https://github.com/Kong/koko/pull/502/files#diff-f39571bb34baebee68e4a03486bb984084067ec6ac049be19f648184888d2494R63-R71

feel free to ping me on on Slack and/or suggest an alternative way to do this. we would also add a readOnly argument to ` newPostgresPool` but not sure this is cleaner anyway.

Reply:
this was discussed offline with Rajan and we agreed it was indeed the simpler way to accomplish what we need in the custom pool.

Reply:
Discussed and the approach is ok